### PR TITLE
[camera] Sample buffer handling on session queue

### PR DESCRIPTION
**Pull request by hellohuanlin**
As discussed in [the proposal](https://docs.google.com/document/d/1Znjp-BduXIb1Ix3JOE1YKyJi1Z0d6rB-2q6Lt7j8xSo/edit#heading=h.kv8nn0uaxog6): 

> SampleBuffer callback on the session queue
> SampleBuffer callback is triggered whenever a new video frame or a new audio sample is arrived. This callback powers live previews, image streaming and video recording. 
> Currently SampleBuffer callback is invoked on either main queue or session queue. It should always be run on the session queue. 

Issue here: https://github.com/flutter/flutter/issues/96429

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  OCMStub([sessionMock canSetSessionPreset:[OCMArg any]]).andReturn(YES);
+
+  dispatch_queue_t captureSessionQueue = dispatch_queue_create("testing", NULL);
+  FLTCam *cam = [[FLTCam alloc] initWithCameraName:@"camera"
+                                  resolutionPreset:@"medium"
+                                       enableAudio:true
+                                       orientation:UIDeviceOrientationPortrait
+                               captureSessionQueue:captureSessionQueue
+                                             error:nil];
+  XCTAssertEqual(captureSessionQueue, cam.captureVideoOutput.sampleBufferCallbackQueue);
```
Comment (jmagman) [inline packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:34]: To use `isEqual:` and not `==`.
```suggestion
  XCTAssertEqualObjects(captureSessionQueue, cam.captureVideoOutput.sampleBufferCallbackQueue);
```

File: packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+@implementation SampleBufferQueueTests
+
+- (void)testSampleBufferCallbackQueueMustBeCaptureSessionQueue {
+  id inputMock = OCMClassMock([AVCaptureDeviceInput class]);
+  OCMStub([inputMock deviceInputWithDevice:[OCMArg any] error:[OCMArg setTo:nil]])
+      .andReturn(inputMock);
+
+  id sessionMock = OCMClassMock([AVCaptureSession class]);
+  OCMStub([sessionMock alloc]).andReturn(sessionMock);
```
Comment (jmagman) [inline packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:23]: If you stub a class method you need to call `-stopMock` explicitly or the mock will leak between tests, it won't stop mocking when the mock object deallocs.

**Review by jmagman — APPROVED** [review]
LGTM with nits.

Can you also do some manual testing of these before you merge?  Want some validation before customers pick up the new version.

File: packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m (hunk 1 above)
Comment (jmagman) [reply packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:34]: Never mind these are `dispatch_queues` not `NSObject`s.

Comment (hellohuanlin) [reply packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:34]: iirc a few years ago apple made dispatch_queues objc objects. So both should be fine. Most likely `isEqual` will simply call `==` internally

Comment (hellohuanlin) [reply packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:34]: found this flag to turn on/off: `OS_OBJECT_HAVE_OBJC_SUPPORT`

File: packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m (hunk 2 above)
Comment (hellohuanlin) [reply packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:23]: gotcha. good to know!

Comment (hellohuanlin) [reply packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:23]: Ah looks like `stopMocking` is called on dealloc, from the doc: https://ocmock.org/reference/

> The class can be returned to its original state by calling stopMocking. This is only necessary if the original state must be restored before the end of the test. The mock automatically calls stopMocking during its own deallocation.

Comment (hellohuanlin) [reply packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:23]: Wrote a dummy test and verified it.

**Comment by hellohuanlin** [comment]
> Can you also do some manual testing of these before you merge? Want some validation before customers pick up the new version.

Sounds good. Just manual tested the sample project and verified its still working.

File: packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m (hunk 2 above)
Comment (jmagman) [reply packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:23]: I didn't think it works stubbing _class_ methods, `alloc` in this case.  Is that what you tested?

Comment (hellohuanlin) [reply packages/camera/camera/example/ios/RunnerTests/SampleBufferQueueTests.m:23]: Yep, i tested `alloc`. 
Screen Shot 2022-01-28 at 2 28 23 PM

**Pull request merged**